File: blockchain.js
// blockchain.js
import crypto from 'node:crypto';

export class NonFungiblePriorityToken {
    constructor(tokenId, ownerTeam, attributes, graphicUri) {
        this.tokenId = tokenId;
        this.ownerTeam = ownerTeam;
        this.timestamp = Date.now();
        this.attributes = attributes;
        this.graphicUri = graphicUri;
        this.signature = this.mintSignatureHash();
    }

    mintSignatureHash() {
        const content = JSON.stringify({
            id: this.tokenId, owner: this.ownerTeam, uri: this.graphicUri, attr: this.attributes
        });
        return crypto.createHash('sha256').update(content).digest('hex');
    }
}

export class Block {
    constructor(index, previousHash, timestamp, data, nftToken = null) {
        this.index = index;
        this.previousHash = previousHash;
        this.timestamp = timestamp;
        this.data = data;
        this.nft = nftToken;
        this.hash = this.computeHash();
    }

    computeHash() {
        const nftSig = this.nft ? this.nft.signature : "NO_NFT";
        const content = JSON.stringify({
            idx: this.index, prev: this.previousHash, payload: this.data, nft: nftSig
        });
        return crypto.createHash('sha256').update(content).digest('hex');
    }
}

export class WyvernNFTBlockchainBackbone {
    constructor() {
        this.chain = [];
        this.nftRegistry = new Map();
        this.createGenesisBlock();
    }

    createGenesisBlock() {
        this.chain.push(new Block(0, "0".repeat(64), Date.now(), { meta: "WYVERN BACKBONE ROOT UNIFIED" }));
    }

    mintPriorityNft(step, coordinates, priorityScore, battery, ownerTeam, graphicUri) {
        const tokenId = `TOKEN-${this.nftRegistry.size + 1}`;
        const attributes = {
            stepIndex: step, geospatialNode: coordinates,
            priorityCoefficient: Number(priorityScore.toFixed(4)), stateOfChargePct: Number(battery.toFixed(2))
        };
        const nft = new NonFungiblePriorityToken(tokenId, ownerTeam, attributes, graphicUri);
        this.nftRegistry.set(tokenId, nft);
        return nft;
    }

    appendSecureBlock(step, coordinates, matrixSlice, metadata, nftToken) {
        const prevBlock = this.chain[this.chain.length - 1];
        const newBlock = new Block(
            this.chain.length, prevBlock.hash, Date.now(),
            { step, gps: coordinates, matrix: matrixSlice, telemetry: metadata }, nftToken
        );
        this.chain.push(newBlock);
        return newBlock.hash.slice(0, 8).toUpperCase();
    }

    verifyLedgerIntegrity() {
        const seenSigs = new Set();
        for (let i = 1; i < this.chain.length; i++) {
            const current = this.chain[i];
            const previous = this.chain[i - 1];

            if (current.hash !== current.computeHash()) return false;
            if (current.previousHash !== previous.hash) return false;

            if (current.nft) {
                if (seenSigs.has(current.nft.signature)) return false;
                seenSigs.add(current.nft.signature);
            }
        }
        return true;
    }
}

File: blockchain.py
import json
import hashlib
import time

class NonFungiblePriorityToken:
    """Represents a unique, serialized asset tracking an operational priority claim."""
    def __init__(self, token_id: str, owner_team: str, attributes: dict, graphic_uri: str):
        self.token_id = token_id
        self.owner_team = owner_team
        self.timestamp = time.time()
        self.attributes = attributes
        self.graphic_uri = graphic_uri
        self.signature = self.mint_signature_hash()

    def mint_signature_hash(self) -> str:
        token_content = json.dumps({
            "id": self.token_id, "owner": self.owner_team, 
            "uri": self.graphic_uri, "attr": self.attributes
        }, sort_keys=True)
        return hashlib.sha256(token_content.encode('utf-8')).hexdigest()

class Block:
    """Represents a secure container block appended to the blockchain ledger."""
    def __init__(self, index: int, previous_hash: str, timestamp: float, data: dict, nft_token=None):
        self.index = index
        self.previous_hash = previous_hash
        self.timestamp = timestamp
        self.data = data
        self.nft = nft_token
        self.hash = self.compute_hash()

    def compute_hash(self) -> str:
        nft_sig = self.nft.signature if self.nft else "NO_NFT"
        block_content = json.dumps({
            "idx": self.index, "prev": self.previous_hash, 
            "payload": self.data, "nft": nft_sig
        }, sort_keys=True)
        return hashlib.sha256(block_content.encode('utf-8')).hexdigest()

class WyvernNFTBlockchainBackbone:
    """Manages secure blocks and enforces ledger consensus rules."""
    def __init__(self):
        self.chain = []
        self.nft_registry = {}
        self.create_genesis_block()

    def create_genesis_block(self):
        self.chain.append(Block(0, "0"*64, time.time(), {"meta": "WYVERN BACKBONE ROOT UNIFIED"}))

    def mint_priority_nft(self, step: int, coordinates: str, priority_score: float, 
                          battery: float, owner_team: str, graphic_uri: str) -> NonFungiblePriorityToken:
        token_id = f"TOKEN-{len(self.nft_registry) + 1}"
        attributes = {
            "step": step, "gps": coordinates, 
            "priority_coeff": round(priority_score, 4), "soc_pct": round(battery, 2)
        }
        nft = NonFungiblePriorityToken(token_id, owner_team, attributes, graphic_uri)
        self.nft_registry[token_id] = nft
        return nft

    def append_secure_block(self, step: int, coordinates: str, matrix_slice: list, 
                            metadata: dict, nft_token: NonFungiblePriorityToken) -> str:
        prev_block = self.chain[-1]
        new_block = Block(
            len(self.chain), prev_block.hash, time.time(), 
            {"step": step, "gps": coordinates, "telemetry": metadata}, nft_token=nft_token
        )
        self.chain.append(new_block)
        return new_block.hash[:8].upper()
